### DISCUSSION TITLE
Use Essentia moods and contexts for tagging music metadata

### DISCUSSION THREAD

**Discussion by Tstassin** [perminder-klair/subwave#543] [Ideas]
Long story short, playing with shows and listening to them at the same time.
Most of the time everything ok with the DJ'ing
At some point, on a focused moment, the DJ chose [Bucephallus Bouncing Ball from Aphex Twin](www.youtube.com/watch?v=cQdhzQkddCc)
I love this song, but it was WAY out of what could be reasonably played at that moment/definition/show.
I took a moment to see all metadata available to Subwave's DJ, and just by looking in the observatory I felt that (even with tts-heavy sidecar) there were objective indices missing telling the truth about the whole engagement needed to appreciate the song.

<img width="355" height="944" alt="Capture d’écran 2026-06-23 à 12 53 22" src="https://github.com/user-attachments/assets/bb4aa342-0211-43d2-8017-78c9a3796624" />

This made me wonder, how could I more "objectively" filter this kind of things at some moment ? Or the inverse, how could I favourite the tendency to IDM/brain intensive songs ?

Essentia provides (subsetting the whole Essentia proposal) [moods and contexts](https://essentia.upf.edu/models.html#moods-and-context)

And by playing with essentia I found (at least) one correct metric to characterize this song : it scores 100/100 in "engagement".
Engagement is defined as 

> Music engagement predicts whether the music evokes active attention of the listener (high-engagement “lean forward” active listening vs. low-engagement “lean back” background listening). 

Essentia providex other really interesting moods / contexts : arousal / danceability / agressivity

Which are triggers that, if integrated, would add a lot to help shape the picking of songs.

If I want to define an electronic focused workable moment I would want to express "be careful of having low agressivity and max medium engagement"
This is about getting mood and contexts into the available metadata (step 1)
Step 2 would be to see how to use / hook this into shows / Djing / picker.

My intro example doesnt mean that I want to fix the problem for this particular song, but its not the first time that the DJ writes that he chose something "same vibe, same energy" but that in reality the follow-up song is nowhere close / mixable and is instant "mood killer/stop the listening".

And I would love to listen to a 1-hour show of agressive/high-engagement EDM, but not at lunch-time with strangers in the room ;-)

**Comment by perminder-klair** [comment]
Good catch, and I think you've nailed the real problem. The DJ says "same energy" but has no objective way to back that up, so now and then it talks itself into a track that's wrong for the room. Bucephalus at lunchtime is a perfect example.

Nice thing is, it fits machinery we already have. There's an analysis pass (`music/analyzer.ts` plus the tts-heavy sidecar) that already decodes each track and runs librosa/Demucs/CLAP on it. Essentia's mood heads take that same audio and return 0 to 1 scores, so Step 1 is mostly adding another model to a worker that's already doing this, plus a couple of DB columns. I'd keep it in the tts-heavy profile since it pulls in TensorFlow.

Step 2 is the interesting half. The key question is soft vs hard. If we just hand the scores to the LLM we're back to trusting "same energy", so for your case the real fix is a hard filter in the picker so an out-of-band track never reaches the model. Then the show defines the rule (something like "low aggression, max-medium engagement"), since shows are already a first-class thing.

Worth noting Essentia isn't redundant with our CLAP embeddings. You can't write "engagement <= medium" against a 512-dim vector, but you can against a named score. That's what makes operator-authored show rules possible.

One caveat: some of the UPF pretrained models are CC BY-NC-SA, so I'd want to check licensing before bundling weights into a published image.

Happy to take a stab at Step 1 if you want to start there.

**Comment by Tstassin** [comment]
Thanks for taking the time to consider this,

While you are at it, I feel like those could help too :

> Approachability, Danceability, Mood Happy, Mood Party, Mood Relaxed, Mood Sad

The more (objective, orthogonal) tuning knobs we get for shows, the more quality can be retrieved for pool candidates / internal steps, and I think this addition is one step further in the flat characterization of the library.

Cheers

**Comment by sxntaxis** [comment]
idk if this goes here but Essentia is capable of tagging multiple genres to a single song (which is realistic, most music doesn't fit into a single genre). But as I understand SUB/WAVE only handles 1 genre per song, and it must be an exact match (meaning that if use multiple genres like "Eletronic;House;Tribal House" the string must exactly match or it won't be recognized). 

Multiple genres allows for more nuances mixes and better accuracy that the generalist approach of single genre songs.

**Reply by Tstassin** [reply]
Totally agree, I'm in the process of planning a retagging of whole 70K+ songs collection with multi-genre (MusicBrainz via beets.io being the source), and multi-genre support is a strong requirement to correctly define songs.
Atm subwave uses only one genre, which implies a scattering of niche genres to correctly define songs, but makes "genre" shows impossible (minimal vs deep house vs tech house vs electronica vs folktronica vs idm vs acid vs ..)
With multi-genres correctly defined on each song, I would love to have multi-genre support and craft shows with filters like "70%-100% genre_1, 40%-70% genre_2, 0%-20% genre_3" this would bring a new world of fine pooling.
